lib/compiler/passes/generate-js.js

```diff
@@ -504,35 +557,42 @@ function generateJS(ast, options) {
             compileCondition("input.length > peg$currPos", 0);
             break;
 
-          case op.MATCH_STRING:       // MATCH_STRING s, a, f, ...
-            compileCondition(
-              ast.literals[bc[ip + 1]].length > 1
-                ? "input.substr(peg$currPos, "
-                    + ast.literals[bc[ip + 1]].length
-                    + ") === "
-                    + l(bc[ip + 1])
-                : "input.charCodeAt(peg$currPos) === "
-                    + ast.literals[bc[ip + 1]].charCodeAt(0),
-              1
+          case op.MATCH_STRING: {     // MATCH_STRING s, a, f, ...
+            const litNum = bc[ip + 1];
+            const literal = ast.literals[litNum];
+            compileInputChunkCondition(
+              inputChunk => {
+                if (literal.length > 1) {
+                  return `${inputChunk} === ${l(litNum)}`;
+                }
+                inputChunk = inputChunk.includes(".charAt(")
+                  ? inputChunk.replace(".charAt(", ".charCodeAt(")
+                  : `${inputChunk}.charCodeAt(0)`;
+                return `${inputChunk} === ${literal.charCodeAt(0)}`;
```

Comment:
It seems, we could always generate `inputChunk` as containing `.charCodeAt` and this hackish replace wouldn't be needed? Or this is not the case for ignored case literals? Then you can pass a `ignoreCase` parameter to `compileInputChunkCondition` and generate different code depending on it.

Or pass in a parameter either `"input.charAt(peg$currPos)"` or `"input.charCodeAt(peg$currPos)"` if this will be more readable (it will be used only if literal has length of 1)

Reply:
The point is that if we use the `inputChunk` twice, we want to do
```
s0 = input.charAt(peg$currPos);
if (s0.charCodeAt(0)) {
  ...
}
```
but if we only use it once, we want to do:
```
if (input.charCodeAt(peg$currPos)) {
  ...
}
```
rather than
```
if (input.charAt(peg$currPos).charCodeAt(0)) {
  ...
}
```

Currently, we never generate byte code that uses the input twice when we're extracting a single, case sensitive character; but thats because there's already an optimization in generate-bytecode to avoid using it twice whenever possible as an optimization.

The way I originally wrote this, with a template string, meant that the code that chose inputChunk, and the code that matched the `charCodeAt(0);` was all self contained in `compileInputChunk`. Now `compileInputChunk` chooses inputChunk, but the callback has to make the decision about charAt vs charCodeAt, so yes, I agree it all looks a bit magical...

I could pass a flag back to the callback saying whether the optimization had happened or not, so the callback could then choose to ignore inputChunk when the optimization didn't happen. Would that work?